nvflare/fuel/utils/config.py

```diff
@@ -0,0 +1,134 @@
+
+from abc import ABC, abstractmethod
+from collections import OrderedDict
+from enum import Enum
+from typing import Dict, Optional
+
+
+class ConfigFormat(Enum):
+    JSON = "JSON"
+    PYHOCON = "PYHOCON"
+    OMEGACONF = "OMEGACONF"
+
+    @classmethod
+    def config_ext_formats(cls):
+        return OrderedDict(
+            {
+                ".json": ConfigFormat.JSON,
+                ".conf": ConfigFormat.PYHOCON,
+                ".yml": ConfigFormat.OMEGACONF,
+                ".json.default": ConfigFormat.JSON,
+                ".conf.default": ConfigFormat.PYHOCON,
+                ".yml.default": ConfigFormat.OMEGACONF,
+            }
+        )
+
+
+class Config(ABC):
+    @abstractmethod
+    def get_native_conf(self):
+        """
+        Return the original underline config object representation if you prefer to use it directly
+        Pyhocon → ConfigTree
+        JSON → Dict
+        OMEGACONF → ConfigDict
+
+        Returns: Any, native config objects
+        """
+
+        pass
+
+    @abstractmethod
+    def get_format(self):
+        """
+        Returns: ConfigFormat, returns the current config objects ConfigFormat
+        """
+
+        pass
+
+    @abstractmethod
+    def get_location(self) -> Optional[str]:
+        """
+        Returns: None if the config is not from file else return the file path where this configuration is loaded from
+
+        """
+        pass
+
+    @abstractmethod
+    def to_dict(self) -> Dict:
+        """
+        convert underline config object to dictionary
+        Returns:Dict
+
+        """
+        pass
+
+    @abstractmethod
+    def to_conf(self, element: Dict) -> str:
+        """
+            convert dict element to the str representation of the underline configuration.
+            For example, for JsonFormat, the method return json string
+            for PyhoconFormat, the method return pyhocon string
+            for OmegaconfFormat, the method returns YAML string representation
+        Args:
+            element: dict
+
+        Returns: str string representation of the configuration in given format
+
+        """
+        pass
+
+
+class ConfigLoader(ABC):
+    @abstractmethod
+    def load_config(
+        self, file_path: str, default_file_path: Optional[str] = None, overwrite_config: Optional[Dict] = None
+    ) -> Config:
+        """
+        configuration from default_file_path will be the default config if specified
+        configuration from file_path will be the merge with default config overwrite the same key
+        configuration from overwrite_config if provided will be the merge with config overwrite the same key
+        Args:
+            file_path: file path for configuration to be loaded
+            default_file_path: file path for default configuration
+            overwrite_config: dict config that will  overwrite the final config if provided
+
+        Returns:nvflare.fuel.utils.Config
+        """
+
+        pass
```

Comment:
Another approach:
What if we have "ConfigLoader" just load the config, so `def load_config(self, file_path)`.
And we let `ConfigFactory` to determine what config should overwrite the other?